src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/NavigableTextAccessibility.m

```diff
@@ -0,0 +1,320 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#import "NavigableTextAccessibility.h"
+#import "JavaAccessibilityUtilities.h"
+#import "ThreadUtilities.h"
+#import "JNIUtilities.h"
+
+static jclass sjc_CAccessibility = NULL;
+#define GET_CACCESSIBLITY_CLASS() \
+     GET_CLASS(sjc_CAccessibility, "sun/lwawt/macosx/CAccessibility");
+#define GET_CACCESSIBLITY_CLASS_RETURN(ret) \
+     GET_CLASS_RETURN(sjc_CAccessibility, "sun/lwawt/macosx/CAccessibility", ret);
+
+static jmethodID sjm_getAccessibleText = NULL;
+#define GET_ACCESSIBLETEXT_METHOD_RETURN(ret) \
+    GET_CACCESSIBLITY_CLASS_RETURN(ret); \
+    GET_STATIC_METHOD_RETURN(sjm_getAccessibleText, sjc_CAccessibility, "getAccessibleText", \
+              "(Ljavax/accessibility/Accessible;Ljava/awt/Component;)Ljavax/accessibility/AccessibleText;", ret);
+
+static jclass sjc_CAccessibleText = NULL;
+#define GET_CACCESSIBLETEXT_CLASS() \
+    GET_CLASS(sjc_CAccessibleText, "sun/lwawt/macosx/CAccessibleText");
+#define GET_CACCESSIBLETEXT_CLASS_RETURN(ret) \
+    GET_CLASS_RETURN(sjc_CAccessibleText, "sun/lwawt/macosx/CAccessibleText", ret);
+
+static jmethodID sjm_getAccessibleEditableText = NULL;
+#define GET_ACCESSIBLEEDITABLETEXT_METHOD_RETURN(ret) \
+    GET_CACCESSIBLETEXT_CLASS_RETURN(ret); \
+    GET_STATIC_METHOD_RETURN(sjm_getAccessibleEditableText, sjc_CAccessibleText, "getAccessibleEditableText", \
+              "(Ljavax/accessibility/Accessible;Ljava/awt/Component;)Ljavax/accessibility/AccessibleEditableText;", ret);
+
+
+@implementation NavigableTextAccessibility
+
+- (BOOL)accessibleIsPasswordText {
+    return [fJavaRole isEqualToString:@"passwordtext"];
+}
+
+// NSAccessibilityElement protocol methods
+
+- (NSRect)accessibilityFrameForRange:(NSRange)range
+{
+    JNIEnv *env = [ThreadUtilities getJNIEnv];
+    GET_CACCESSIBLETEXT_CLASS_RETURN(NSMakeRect(0, 0, 0, 0));
+    DECLARE_STATIC_METHOD_RETURN(jm_getBoundsForRange, sjc_CAccessibleText, "getBoundsForRange",
+                         "(Ljavax/accessibility/Accessible;Ljava/awt/Component;II)[D", NSMakeRect(0, 0, 0, 0));
+    jdoubleArray axBounds = (jdoubleArray)(*env)->CallStaticObjectMethod(env, sjc_CAccessibleText, jm_getBoundsForRange,
+                              fAccessible, fComponent, range.location, range.length);
+    CHECK_EXCEPTION();
+    if (axBounds == NULL) return NSMakeRect(0, 0, 0, 0);
+
+    // We cheat because we know that the array is 4 elements long (x, y, width, height)
+    jdouble *values = (*env)->GetDoubleArrayElements(env, axBounds, 0);
+    CHECK_EXCEPTION();
+    if (values == NULL) {
+        // Note: Java will not be on the stack here so a java exception can't happen and no need to call ExceptionCheck.
+        NSLog(@"%s failed calling GetDoubleArrayElements", __FUNCTION__);
```

Comment:
Please clarify the comment above.

Reply:
This comment has been carried over from the legacy code. it is still relevant.

Reply:
The comment states that it is not necessary to call ExceptionCheck, but it will be done before by the CHECK_EXCEPTION. The macro also will log it when necessary.

Reply:
I suggest doing the opposite, leave the CHECK_EXCEPTION and remove the chech and the logging. The CHECK_EXCEPTION  is better since it can be configured via properties, and it will chech itself on what thread the current code is executed.

Reply:
Done